pretyping/coercion.ml

```diff
@@ -582,19 +584,68 @@ let inh_coerce_to_base ?loc ~program_mode env sigma j =
+  if c1 = c2 then []
+  else lookup_path_between_class (c1,c2)
+
+(* We find a path to a common class such that the path from src_expected is
+   reversible (the casted term is unknown but inferrable via unification,
+   eg CS inference) *)
+let lookup_reversible_path_to_common_point env sigma ~src_expected ~src_inferred =
+  let _, c_expected = class_of env sigma src_expected in
+  let _, c_inferred = class_of env sigma src_inferred in
+  let reachable_from_expected = reachable_from c_expected in
+  let reachable_from_inferred = reachable_from c_inferred in
+  let r = ClTypSet.(elements (inter reachable_from_expected reachable_from_inferred)) in
+  (* XXX this order relation is not total ... *)
+  let r = List.sort (fun c1 c2 -> if ClTypSet.mem c1 (reachable_from c2) then 1 else -1) r in
+  let r = (if ClTypSet.mem c_inferred reachable_from_expected then [c_inferred] else []) @ r in
+  (* This is unneeded since there is a dedicated case in inh_coerce_to_fail:
+     let r = (if ClTypSet.mem c_expected reachable_from_inferred then [c_expected] else []) @ r in *)
+  let rec aux = function
+  | [] -> raise Not_found
+  | c :: cs ->
+      let reversible = lookup_path_between_class_reflexive (c_expected,c) in
+      if List.for_all has_reverse_attribute reversible then
+        let direct = lookup_path_between_class_reflexive (c_inferred,c) in
+        class_has_args c_expected, class_has_args c_inferred, reversible, direct
+      else
+        aux cs
+  in
+    aux r
+
 let inh_coerce_to_fail flags env sigma rigidonly v v_ty target_type =
   if rigidonly && not (Heads.is_rigid env (EConstr.Unsafe.to_constr target_type) && Heads.is_rigid env (EConstr.Unsafe.to_constr v_ty))
   then
     raise NoCoercion
   else
-    let sigma, v', v'_ty, trace =
-      try
-        let c = lookup_path_between env sigma ~src:v_ty ~tgt:target_type in
-        apply_coercion env sigma c v v_ty
-      with Not_found -> raise NoCoercion
-    in
-    try unify_leq_delay ~flags env sigma v'_ty target_type, v', trace
-    with Evarconv.UnableToUnify _ as exn ->
+    try
+      let sigma, v', v'_ty, trace =
+        try
+          (* 1 path c of direct coercions *)
+          let c = lookup_path_between env sigma ~src:v_ty ~tgt:target_type in
+          apply_coercion env sigma c v v_ty
+        with Not_found ->
+          (* 2 paths: one of direct coercions, one of reversible coercions *)
+          let target_type_has_args, v_ty_has_args, reversible, direct =
+            lookup_reversible_path_to_common_point env sigma ~src_expected:target_type ~src_inferred:v_ty in
+          if not (v_ty_has_args || target_type_has_args) then raise Not_found;
+          let sigma, x = Evarutil.new_evar env sigma target_type in
+          let sigma, rev_x, _, _ = apply_coercion env sigma reversible x target_type in
+          let sigma, direct_v, _, _ = apply_coercion env sigma direct v v_ty in
+          let sigma = unify_leq_delay ~flags env sigma direct_v rev_x in
```

Comment:
Maybe use the non-delay unify instead?

Reply:
Wouldn't that fail if there are remaining evars? (that would make the new reverse_coercion_ac.v test in this PR fail)

Reply:
Try and see

Reply:
Tried, it's the exact contrary of what I feared, it doesn't fail when `x` remains uninstantiated.
Thanks for the suggestion anyway.